dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/resources_plugin/gitlab.py

```diff
@@ -0,0 +1,118 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""DolphinScheduler gitlab resource plugin."""
+from typing import Optional
+from urllib.parse import urljoin
+
+import gitlab
+import requests
+
+from pydolphinscheduler.core.resource_plugin import ResourcePlugin
+from pydolphinscheduler.resources_plugin.base.git import Git, GitLabFileInfo
+
+
+class GitLab(ResourcePlugin, Git):
+    """GitLab object, declare GitLab resource plugin for task and workflow to dolphinscheduler.
+
+    :param prefix: A string representing the prefix of GitLab.
+    :param private_token: A string used for identity authentication of GitLab private or Internal warehouse.
+    :param oauth_token: A string used for identity authentication of GitLab private or Internal warehouse.
+    :param username: A string representing the user of the warehouse.
+    :param password: A string representing the user password.
+    """
+
+    def __init__(
+        self,
+        prefix: str,
+        private_token: Optional[str] = None,
+        oauth_token: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        *args,
+        **kwargs
+    ):
+        super().__init__(prefix, *args, **kwargs)
+        self.private_token = private_token
+        self.oauth_token = oauth_token
+        self.username = username
+        self.password = password
+
+    def get_git_file_info(self, path: str):
+        """Get file information from the file url, like repository name, user, branch, and file path."""
+        elements = path.split("/")
+        self.get_index(path, "/", 8)
+        for i in range(0, len(elements)):
+            if (
+                i + 3 < len(elements)
+                and elements[i + 1] == "-"
+                and elements[i + 2] == "blob"
+            ):
+                host_end = self.get_index(path, "/", 3)
+                self._git_file_info = GitLabFileInfo(
+                    host=path[0:host_end],
+                    repo_name=elements[i],
+                    branch=elements[i + 3],
+                    file_path="/".join(
+                        str(elements[j]) for j in range(i + 4, len(elements))
+                    ),
+                    user="/".join(str(elements[j]) for j in range(3, i)),
+                )
```

Comment:
can we use other way to handle this, I think we use too many constant in this method

Reply:
after we discuss `url.parse` can not handle this issue.